NullPointerExcetion or ArrayIndexOutOfBoundsException in SearchBridge when using custom input/output configuration.

**Describe the bug**  
When running with federatedRegistries-1.0, it is possible to get an NPE (when the output attribute is single-valued) or ArrayIndexOutOfBoundsException (when the output attribute is multi-valued) when defining a non-identifier type property for the federated registries group or user output security name mapping when federating a UserRegistry (such as BasicRegistry or SafRegistry) as they don't support non-identifier properties. These exceptions occur in the `com.ibm.ws.security.wim.registry.util.SearchBridge` class. This is an example exception:


```
com.ibm.websphere.security.CustomRegistryException
	at com.ibm.ws.security.registry.internal.UserRegistryWrapper.getUsers(UserRegistryWrapper.java:101)
	at web.UserRegistryServlet.handleMethodRequest(UserRegistryServlet.java:99)
	at web.UserRegistryServlet.doGet(UserRegistryServlet.java:174)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1258)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:746)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:443)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1227)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1011)
	at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:75)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:938)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:281)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1246)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:468)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:427)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:567)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:501)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:361)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:328)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:167)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:75)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:514)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:584)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:968)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1057)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:245)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:836)
Caused by: com.ibm.ws.security.registry.RegistryException
	at com.ibm.ws.security.wim.registry.WIMUserRegistry.getUsers(WIMUserRegistry.java:258)
	at com.ibm.ws.security.registry.internal.UserRegistryWrapper.getUsers(UserRegistryWrapper.java:92)
	... 29 more
Caused by: java.lang.NullPointerException
	at com.ibm.ws.security.wim.registry.util.SearchBridge.getUsers(SearchBridge.java:167)
	at com.ibm.ws.security.wim.registry.WIMUserRegistry.getUsers(WIMUserRegistry.java:252)
	... 30 more
```


**Steps to Reproduce**  
Configure an LDAP registry and a basic registry (with at least one user). Configure the userSecurityNameMapping input property to a non-identifier type property, such as mail below.

```
    <basicRegistry realm="BasicRealm">
      <user name="user1" password="mypassword" />
    </basicRegistry>

    <ldapRegistry realm="LdapRealm"... >
        ...
    </ldapRegistry>

    <federatedRepository>
      <primaryRealm name="FederatedRealm">
	<participatingBaseEntry name="dc=example,dc=com"/>
	<participatingBaseEntry name="o=BasicRealm"/> 
	<userSecurityNameMapping inputProperty="mail" outputProperty="mail"/>
      </primaryRealm>
    </federatedRepository>
```

Call the `getUsers` UserRegistry method and you will receive an NPE. The key to reproducing is setting the user or group security name output mapping for the to one that is not an identifier type property (uniqueName, uniqueId, externalName, externalId).

**Expected behavior**  
No NPE or ArrayIndexOutOfBoundsException should be thrown. The code should gracefully handle entities that are returned that don't have the property. Either they should be excluded, the default output attribute should be used instead, or a registry exception should be thrown.

**Diagnostic information:**  
 - OpenLiberty Version: All
 - Affected feature(s): federatedRegistries-1.0
 - Java Version: All
 - server.xml configuration: See above.

**Additional context**  
None

--- a/dev/com.ibm.ws.security.wim.registry/src/com/ibm/ws/security/wim/registry/util/SearchBridge.java
+++ b/dev/com.ibm.ws.security.wim.registry/src/com/ibm/ws/security/wim/registry/util/SearchBridge.java
@@ -1,10 +1,10 @@
 /*******************************************************************************
- * Copyright (c) 2012, 2020 IBM Corporation and others.
+ * Copyright (c) 2012, 2023 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -160,16 +160,9 @@ public class SearchBridge {
                         break;
                     }
                     Entity loginAccount = returnedList.get(count);
-                    // d115256
-                    if (!this.mappingUtils.isIdentifierTypeProperty(outputAttrName)) {
-                        Object value = loginAccount.get(outputAttrName);
-                        if (value instanceof String) {
-                            people.add((String) value);
-                        } else {
-                            people.add(BridgeUtils.getStringValue(((List<?>) value).get(0)));
-                        }
-                    } else {
-                        people.add(BridgeUtils.getStringValue(loginAccount.getIdentifier().get(outputAttrName)));
+                    String sName = getSecurityNameFromEntity(methodName, outputAttrName, loginAccount);
+                    if (sName != null) {
+                        people.add(sName);
                     }
                 }
                 returnValue = new SearchResult(people, true);
@@ -350,15 +343,9 @@ public class SearchBridge {
                     }
                     // d113801
                     if (isEntityTypeGrp) {
-                        if (!this.mappingUtils.isIdentifierTypeProperty(outputAttrName)) {
-                            Object value = group.get(outputAttrName);
-                            if (value instanceof String) {
-                                groups.add((String) value);
-                            } else {
-                                groups.add(BridgeUtils.getStringValue(((List<?>) value).get(0)));
-                            }
-                        } else {
-                            groups.add(BridgeUtils.getStringValue(group.getIdentifier().get(outputAttrName)));
+                        String sName = getSecurityNameFromEntity(methodName, outputAttrName, group);
+                        if (sName != null) {
+                            groups.add(sName);
                         }
                     } else {
                         if (tc.isEventEnabled()) {
@@ -392,5 +379,39 @@ public class SearchBridge {
             }
         }
         return returnValue;
+    }
+
+    /**
+     * Get the security name from the provided entity.
+     *
+     * @param methodName     The name of the method calling.
+     * @param outputAttrName The name of the attribute to use for the security name.
+     * @param entity         The entity to get the value from.
+     * @return The security name representing the entity, or null if one could not be found.
+     */
+    private String getSecurityNameFromEntity(String methodName, String securityNameAttr, Entity entity) {
+        String entityName = null;
+        if (!this.mappingUtils.isIdentifierTypeProperty(securityNameAttr)) {
+            Object value = entity.get(securityNameAttr);
+            if (value instanceof String) {
+                entityName = (String) value;
+            } else if (value instanceof List<?> && !((List<?>) value).isEmpty()) {
+                entityName = BridgeUtils.getStringValue(((List<?>) value).get(0));
+            } else {
+                /*
+                 * The outputAttrName did not exist in the returned entity.
+                 * It will be excluded from results.
+                 */
+                if (tc.isDebugEnabled()) {
+                    Tr.debug(tc, methodName + " Excluding the following entity from results as it does not " +
+                                 "contain the specified output attribute {0}.\n\n{1}",
+                             securityNameAttr, entity);
+                }
+            }
+        } else {
+            entityName = BridgeUtils.getStringValue(entity.getIdentifier().get(securityNameAttr));
+        }
+
+        return entityName;
     }
 }

--- a/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/WIMURRegressionTest.java
+++ b/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/WIMURRegressionTest.java
@@ -1,10 +1,10 @@
 /*******************************************************************************
- * Copyright (c) 2019, 2022 IBM Corporation and others.
+ * Copyright (c) 2019, 2023 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -29,11 +29,13 @@ import com.ibm.websphere.simplicity.config.BasicRegistry;
 import com.ibm.websphere.simplicity.config.BasicRegistry.Group;
 import com.ibm.websphere.simplicity.config.BasicRegistry.Group.Member;
 import com.ibm.websphere.simplicity.config.BasicRegistry.User;
+import com.ibm.websphere.simplicity.config.ConfigElementList;
 import com.ibm.websphere.simplicity.config.ServerConfiguration;
 import com.ibm.websphere.simplicity.config.wim.BaseEntry;
 import com.ibm.websphere.simplicity.config.wim.FederatedRepository;
 import com.ibm.websphere.simplicity.config.wim.LdapRegistry;
 import com.ibm.websphere.simplicity.config.wim.Realm;
+import com.ibm.websphere.simplicity.config.wim.RealmPropertyMapping;
 import com.ibm.websphere.simplicity.log.Log;
 import com.ibm.ws.com.unboundid.InMemoryLDAPServer;
 import com.ibm.ws.security.registry.SearchResult;
@@ -181,6 +183,7 @@ public class WIMURRegressionTest {
         entry.addAttribute("uid", USER);
         entry.addAttribute("sn", USER + "_sn");
         entry.addAttribute("cn", USER + "_cn");
+        entry.addAttribute("mail", USER + "@acme.org");
         entry.addAttribute("userPassword", USER_PASSWORD);
         ldapServer1.add(entry);
 
@@ -195,6 +198,7 @@ public class WIMURRegressionTest {
         entry = new Entry(GROUP_DN);
         entry.addAttribute("objectclass", "groupOfNames");
         entry.addAttribute("cn", GROUP_CN);
+        entry.addAttribute("description", GROUP_CN + "_description");
         entry.addAttribute("member", USER_DN);
         ldapServer1.add(entry);
     }
@@ -410,5 +414,80 @@ public class WIMURRegressionTest {
          */
         assertFalse("Calling isValidUser() with valid group should return false.", servlet.isValidUser(BASIC_GROUP));
         assertTrue("Calling isValidUser() with valid user should return true.", servlet.isValidUser(BASIC_USER));
+    }
+
+    /**
+     * When non-idenfitifier properies are configured as the group and user security name output property
+     * mappings in Liberty, it is possible that they don't exist. In fact, they will NOT exist when the entity
+     * is from a UserRegistry and not a Repository. Make sure we handle those properties missing.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void test_SearchBridge_withNonIdentifierOutputProps() throws Exception {
+        ServerConfiguration clone = startConfiguration.clone();
+
+        /*
+         * Create an LDAP registry.
+         */
+        createLdapRegistry1(clone);
+
+        /*
+         * Create a basic registry.
+         */
+        BasicRegistry registry = new BasicRegistry();
+        registry.setRealm("BasicRealm");
+        User user = new User();
+        user.setName(BASIC_USER);
+        user.setPassword(BASIC_USER_PASSWORD);
+        registry.getUsers().add(user);
+        Group group = new Group();
+        group.setName(BASIC_GROUP);
+        Member member = new Member();
+        member.setName(BASIC_USER);
+        group.getMembers().add(member);
+        registry.getGroups().add(group);
+        clone.getBasicRegistries().add(registry);
+
+        /*
+         * Explicitly configure federated registries.
+         */
+        FederatedRepository federatedRepository = new FederatedRepository();
+        Realm primaryRealm = new Realm();
+        primaryRealm.setName("FederatedRealm");
+        ConfigElementList<BaseEntry> pes = new ConfigElementList<BaseEntry>();
+        pes.add(new BaseEntry("o=BasicRealm"));
+        pes.add(new BaseEntry(LDAP1_BASE_DN));
+        primaryRealm.setParticipatingBaseEntries(pes);
+
+        /*
+         * Group input/output mappings.
+         */
+        primaryRealm.setGroupSecurityNameMapping(new RealmPropertyMapping("description", "description"));
+
+        /*
+         * User input/output mappings.
+         */
+        primaryRealm.setUserSecurityNameMapping(new RealmPropertyMapping("mail", "mail"));
+
+        federatedRepository.setPrimaryRealm(primaryRealm);
+        clone.setFederatedRepositoryElement(federatedRepository);
+
+        LDAPFatUtils.updateConfigDynamically(libertyServer, clone);
+
+        /*
+         *
+         */
+        SearchResult result = servlet.getGroups("*", 10);
+        assertEquals("Expect only the LDAP group in the getGroups() result.", 1, result.getList().size());
+        assertTrue("Expect only the LDAP group in the getGroups() result.", "group_description".equals(result.getList().get(0)));
+
+        /*
+         * SearchBridge.getUsers() would throw an NPE when a UserRegistry user came back because
+         * the output attribute would not exist, unless it was an identifier property.
+         */
+        result = servlet.getUsers("*", 10);
+        assertEquals("Expect only the LDAP user in the getUsers() result.", 1, result.getList().size());
+        assertTrue("Expect only the LDAP user in the getUsers() result.", "test//@acme.org".equals(result.getList().get(0)));
     }
 }

